File: sliding-window-dive/repro.py
"""Reproduction harness for the sliding-window "no valid trim point found" bug (Python).

Runs SlidingWindowConversationManager.reduce_context / apply_management against the SAME
174-message fixture the TypeScript repro serialized (fixtures/messages-174.json), plus the
other trigger shapes. No network, no model: only ``agent.messages`` is touched.

Run:  sliding-window-dive/.venv/bin/python sliding-window-dive/repro.py
"""

import json
import logging
import sys
from pathlib import Path
from types import SimpleNamespace

from strands.agent.conversation_manager.sliding_window_conversation_manager import (
    SlidingWindowConversationManager,
)
from strands.types.exceptions import ContextWindowOverflowException

DIVE_DIR = Path(__file__).parent

# ---------- log capture ----------
captured: list[str] = []


class _Capture(logging.Handler):
    def emit(self, record: logging.LogRecord) -> None:
        captured.append(f"{record.levelname} {record.getMessage()}")


logging.getLogger("strands").addHandler(_Capture())
logging.getLogger("strands").setLevel(logging.DEBUG)

# ---------- history builders (dict Messages, matching strands-py types) ----------
_seq = 0


def user_text(text):
    return {"role": "user", "content": [{"text": text}]}


def assistant_text(text):
    return {"role": "assistant", "content": [{"text": text}]}


def tool_use_msg(with_text=False):
    global _seq
    _seq += 1
    content = []
    if with_text:
        content.append({"text": f"Let me check that (step {_seq})."})
    content.append({"toolUse": {"name": "get_status", "toolUseId": f"tool_{_seq}", "input": {"step": _seq}}})
    return {"role": "assistant", "content": content}


def tool_result_msg():
    return {
        "role": "user",
        "content": [
            {
                "toolResult": {
                    "toolUseId": f"tool_{_seq}",
                    "status": "success",
                    "content": [{"text": f"ok step={_seq}"}],
                }
            }
        ],
    }


def build_s1(total=174):
    """No plain user in tail + toolUse/toolResult never adjacent (interleaved assistant text)."""
    messages = [user_text("Build a cabin near my position and keep me posted.")]
    messages.append(assistant_text("On it — starting the survey."))
    while len(messages) < total - 1:
        messages.append(tool_use_msg())
        messages.append(assistant_text(f"Reading the result of step {_seq}..."))
        messages.append(tool_result_msg())
    del messages[total - 1 :]
    messages.append(assistant_text("Progress update: foundation laid."))
    return messages[:total]


def build_s2(total=174):
    """Adjacent pairs, no plain user anywhere."""
    messages = []
    while len(messages) < total:
        messages.append(tool_use_msg(True))
        messages.append(tool_result_msg())
    return messages[:total]


def build_s7(total=174):
    """Plain agentic: plain user at 0, adjacent pairs after."""
    messages = [user_text("Chop wood until I say stop.")]
    while len(messages) < total:
        messages.append(tool_use_msg())
        messages.append(tool_result_msg())
    return messages[:total]


def load_fixture():
    with open(DIVE_DIR / "fixtures" / "messages-174.json") as f:
        raw = json.load(f)
    return [{"role": m["role"], "content": m["content"]} for m in raw]


# ---------- runner ----------
lines: list[str] = []


def out(s):
    lines.append(s)
    print(s)


def describe(messages):
    plain = sum(
        1
        for m in messages
        if m["role"] == "user" and not any(("toolResult" in c or "toolUse" in c) for c in m["content"])
    )
    return f"len={len(messages)} plainUserMsgs={plain}"


def run_case(name, messages, window_size=120, error=None, rounds=3):
    manager = SlidingWindowConversationManager(window_size=window_size)
    agent = SimpleNamespace(messages=messages)
    out(f"\n## {name}")
    out(f"before: {describe(messages)} window_size={window_size} e={'set' if error else 'None'}")
    for rnd in range(1, rounds + 1):
        captured.clear()
        before = len(messages)
        try:
            manager.reduce_context(agent, e=error)
            reduced = len(messages) < before
            out(f"round {rnd}: reduced={reduced} after: {describe(messages)}")
        except ContextWindowOverflowException as exc:
            out(f"round {rnd}: RAISED ContextWindowOverflowException: {exc}")
        for log_line in captured:
            out(f"  log: {log_line}")
        if len(messages) >= before:
            # emulate one more agent turn while stuck over the window
            messages.append(tool_use_msg())
            messages.append(assistant_text("interleaved"))
            messages.append(tool_result_msg())
    first = messages[0]
    blocks = ",".join(sorted(k for c in first["content"] for k in c))
    out(f"post-trim first message: role={first['role']} blocks=[{blocks}] | over-window={len(messages) > window_size}")


out("# repro.py — SlidingWindowConversationManager, window=120")

run_case("S1 FIXTURE (TS-serialized real crash shape, e=None routine management)", load_fixture())
run_case("S1 FIXTURE with e set (reactive overflow recovery)", load_fixture(), error=RuntimeError("input too long"), rounds=1)
run_case("S2 adjacent pairs, no plain user anywhere (PY fallback trims w/o anchor)", build_s2())
run_case("S7 plain agentic shape", build_s7())

with open(DIVE_DIR / (sys.argv[1] if len(sys.argv) > 1 else "repro-output-py.txt"), "w") as f:
    f.write("\n".join(lines) + "\n")

File: sliding-window-dive/repro.ts
/**
 * Reproduction harness for the sliding-window "no valid trim point found" bug (TypeScript).
 *
 * Runs SlidingWindowConversationManager.reduce() (the same path AfterInvocationEvent's
 * _applyManagement takes) against realistic tool-heavy histories. No network, no model:
 * `reduce` only touches `agent.messages`.
 *
 * Run from repo root:  node_modules/.bin/tsx sliding-window-dive/repro.ts
 */
import { mkdirSync, writeFileSync } from 'node:fs'
import { dirname, join } from 'node:path'
import { fileURLToPath } from 'node:url'

import { Message, TextBlock, ToolUseBlock, ToolResultBlock } from '../strands-ts/src/types/messages.js'
import { SlidingWindowConversationManager } from '../strands-ts/src/conversation-manager/sliding-window-conversation-manager.js'
import { configureLogging } from '../strands-ts/src/logging/logger.js'
import type { LocalAgent } from '../strands-ts/src/types/agent.js'
import type { Model } from '../strands-ts/src/models/model.js'

const DIVE_DIR = dirname(fileURLToPath(import.meta.url))

// ---------- log capture ----------
const captured: string[] = []
configureLogging({
  debug: (...a) => captured.push(`DEBUG ${a.join(' ')}`),
  info: (...a) => captured.push(`INFO ${a.join(' ')}`),
  warn: (...a) => captured.push(`WARN ${a.join(' ')}`),
  error: (...a) => captured.push(`ERROR ${a.join(' ')}`),
})

// ---------- history builders ----------
let toolSeq = 0
function userText(text: string): Message {
  return new Message({ role: 'user', content: [new TextBlock(text)] })
}
function assistantText(text: string): Message {
  return new Message({ role: 'assistant', content: [new TextBlock(text)] })
}
function toolUseMsg(withText = false): Message {
  toolSeq++
  const content: (TextBlock | ToolUseBlock)[] = []
  if (withText) content.push(new TextBlock(`Let me check that (step ${toolSeq}).`))
  content.push(new ToolUseBlock({ name: 'get_status', toolUseId: `tool_${toolSeq}`, input: { step: toolSeq } }))
  return new Message({ role: 'assistant', content })
}
function toolResultMsg(id?: string): Message {
  return new Message({
    role: 'user',
    content: [
      new ToolResultBlock({
        toolUseId: id ?? `tool_${toolSeq}`,
        status: 'success',
        content: [new TextBlock(`ok: position=(133.5, 79, -69.5) health=20 food=20 step=${toolSeq}`)],
      }),
    ],
  })
}

/**
 * S1 — the user's real crash shape: 174 messages, window 120.
 * A long autonomous agentic run where toolUse and its toolResult are separated by an
 * interleaved assistant message (concurrent-fork fold-back / mid-turn injection /
 * separate reasoning message). The 120-message tail therefore contains NO plain user
 * message and NO *adjacent* assistant(toolUse)+user(toolResult) pair.
 */
function buildS1(total = 174): Message[] {
  const messages: Message[] = [userText('Build a cabin near my position and keep me posted.')]
  messages.push(assistantText('On it — starting the survey.'))
  // repeating triple: assistant(toolUse), assistant(text), user(toolResult)
  while (messages.length < total - 1) {
    messages.push(toolUseMsg())
    messages.push(assistantText(`Reading the result of step ${toolSeq}...`))
    messages.push(toolResultMsg())
  }
  while (messages.length > total - 1) messages.pop()
  messages.push(assistantText('Progress update: foundation laid.'))
  return messages.slice(0, total)
}

/** S2 — adjacent pairs exist, but NO plain user message anywhere (restored/imported session). */
function buildS2(total = 174): Message[] {
  const messages: Message[] = []
  while (messages.length < total) {
    messages.push(toolUseMsg(true))
    messages.push(toolResultMsg())
  }
  return messages.slice(0, total)
}

/** S3 — trailing orphaned toolUse; everything else non-adjacent tool traffic. */
function buildS3(total = 174): Message[] {
  const messages = buildS1(total - 1)
  messages.push(toolUseMsg()) // awaiting result — orphan
  return messages
}

/** S4 — pinFirst pins a prefix that is NOT a valid user-first alternating run. */
function buildS4(total = 174): Message[] {
  const messages: Message[] = [assistantText('System bootstrap notice (assistant-first).')]
  messages.push(userText('hi'))
  while (messages.length < total) {
    messages.push(toolUseMsg())
    messages.push(toolResultMsg())
  }
  return messages.slice(0, total)
}

/** S7 — plain agentic shape: plain user at 0, adjacent pairs after; anchor fallback "works". */
function buildS7(total = 174): Message[] {
  const messages: Message[] = [userText('Chop wood until I say stop.')]
  while (messages.length < total) {
    messages.push(toolUseMsg())
    messages.push(toolResultMsg())
  }
  return messages.slice(0, total)
}

// ---------- runner ----------
function describeHistory(messages: Message[]): string {
  const plainUsers = messages.filter(
    (m) => m.role === 'user' && !m.content.some((b) => b.type === 'toolResultBlock' || b.type === 'toolUseBlock')
  ).length
  return `len=${messages.length} plainUserMsgs=${plainUsers}`
}

type Case = { name: string; build: () => Message[]; windowSize?: number; pinFirst?: number }
const cases: Case[] = [
  { name: 'S1 no-plain-user + no-adjacent-pair (REAL CRASH SHAPE)', build: () => buildS1() },
  { name: 'S2 adjacent pairs, no plain user anywhere (anchor missing)', build: () => buildS2() },
  { name: 'S3 S1 + trailing orphaned toolUse', build: () => buildS3() },
  { name: 'S4 pinFirst=2 over assistant-first prefix', build: () => buildS4(), pinFirst: 2 },
  { name: 'S7 plain agentic (anchor works) — off-by-one check', build: () => buildS7() },
]

const lines: string[] = []
const out = (s: string) => {
  lines.push(s)
  console.log(s)
}

out(`# repro.ts — SlidingWindowConversationManager, window=120`)
for (const c of cases) {
  const windowSize = c.windowSize ?? 120
  const messages = c.build()
  const manager = new SlidingWindowConversationManager({
    windowSize,
    ...(c.pinFirst !== undefined ? { pinFirst: c.pinFirst } : {}),
  })
  const agent = { messages } as unknown as LocalAgent
  const model = {} as Model

  out(`\n## ${c.name}`)
  out(`before: ${describeHistory(messages)} windowSize=${windowSize}`)

  // Simulate 3 consecutive invocations: AfterInvocation applies reduce when over the window.
  for (let round = 1; round <= 3; round++) {
    captured.length = 0
    const reduced = manager.reduce({ agent, model } as never)
    out(`round ${round}: reduced=${reduced} after: ${describeHistory(messages)}`)
    for (const l of captured) out(`  log: ${l}`)
    // history keeps growing between invocations in real life; emulate one more turn
    if (reduced === false) {
      messages.push(toolUseMsg())
      messages.push(assistantText('interleaved'))
      messages.push(toolResultMsg())
    }
  }
  const first = messages[0]!
  out(
    `post-trim first message: role=${first.role} blocks=[${first.content.map((b) => b.type).join(',')}] | over-window=${messages.length > windowSize}`
  )
}

// ---------- fixture: serialize the exact real-crash shape ----------
const fixture = buildS1(174)
mkdirSync(join(DIVE_DIR, 'fixtures'), { recursive: true })
writeFileSync(
  join(DIVE_DIR, 'fixtures', 'messages-174.json'),
  JSON.stringify(
    fixture.map((m) => ({ role: m.role, content: m.content.map((b) => b.toJSON()) })),
    null,
    2
  )
)
out(`\nfixture written: fixtures/messages-174.json (${fixture.length} messages)`)

writeFileSync(join(DIVE_DIR, process.argv[2] ?? 'repro-output-ts.txt'), lines.join('\n') + '\n')
